### PR TITLE
System Table to-commit/commit-hash lookups

### PR DESCRIPTION
**Pull request by max-hoffman**
Filters and joins on system tables can take use commit hashes in
equality filters as direct lookups into the node store. System tables
reject non-point lookups.

Missing `from_commit` optimizations.

### CODE REVIEW

**Comment by max-hoffman** [comment]
preliminary results, subject to inaccuracy as I review:

before:
```
+-------------------------------------+-------+-----------+
| name                                | mean  | mean_mult |
+-------------------------------------+-------+-----------+
| dolt_commit_ancestors_commit_filter | 1.99  | 7.96      |
| dolt_commits_commit_filter          | 1.7   | 6.54      |
| dolt_diff_log_join_on_commit        | 50.93 | 1.08      |
| dolt_diff_table_commit_filter       | 3.43  | 8.79      |
| dolt_diffs_commit_filter            | 0.45  | 1.73      |
| dolt_history_commit_filter          | 1.22  | 3.13      |
| dolt_log_commit_filter              | 2.87  | 11.04     |
+-------------------------------------+-------+-----------+
+------+
| avg  |
+------+
| 5.75 |
+------+
```

after:
```
+-------------------------------------+-------+-----------+
| name                                | mean  | mean_mult |
+-------------------------------------+-------+-----------+
| dolt_commit_ancestors_commit_filter | 0.25  | 0.93      |
| dolt_commits_commit_filter          | 0.33  | 1.22      |
| dolt_diff_log_join_on_commit        | 87.31 | 1.63      |
| dolt_diff_table_commit_filter       | 0.58  | 1.49      |
| dolt_diffs_commit_filter            | 0.33  | 1.22      |
| dolt_history_commit_filter          | 0.48  | 1.17      |
| dolt_log_commit_filter              | 0.3   | 1.11      |
+-------------------------------------+-------+-----------+
+------+
| avg  |
+------+
| 1.25 |
+------+
```

File: go/libraries/doltcore/sqle/enginetest/dolt_queries_diff.go [hunk 1]
```diff
@@ -2703,3 +2703,165 @@ var CommitDiffSystemTableScriptTests = []queries.ScriptTest{
 		},
 	},
 }
+
+type systabScript struct {
+	name     string
+	setup    []string
+	userVars []string
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/enginetest/dolt_queries_diff.go:2710]: Why do these need to be separate from setup?

File: go/libraries/doltcore/sqle/history_table.go [hunk 2]
```diff
@@ -76,14 +77,62 @@ func (ht *HistoryTable) GetIndexes(ctx *sql.Context) ([]sql.Index, error) {
...
+
+			}
+			h, ok := hash.MaybeParse(hs)
+			if !ok {
+				continue
+			}
+			hashes = append(hashes, h)
+
+			cm, err := doltdb.HashToCommit(ctx, ht.doltTable.db.DbData().Ddb.ValueReadWriter(), ht.doltTable.db.DbData().Ddb.NodeStore(), h)
+			if err != nil {
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/history_table.go:114]: Same comment

File: go/libraries/doltcore/sqle/tables.go [hunk 3]
```diff
@@ -183,8 +183,8 @@ var _ doltReadOnlyTableInterface = (*DoltTable)(nil)
 //var _ sql.ProjectedTable = (*DoltTable)(nil)
 
 // IndexedAccess implements sql.IndexAddressableTable
-func (t *DoltTable) IndexedAccess(idx sql.Index) sql.IndexedTable {
-	return NewIndexedDoltTable(t, idx.(index.DoltIndex))
+func (t *DoltTable) IndexedAccess(lookup sql.IndexLookup) sql.IndexedTable {
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/tables.go:186]: This seems bad... were we not getting indexes before this change? How did this even work?

**Review by zachmu — APPROVED** [review]
LGTM, mostly the same comment multiple times

File: go/libraries/doltcore/sqle/dtables/commits_table.go [hunk 4]
```diff
@@ -69,8 +73,84 @@ func (dt *CommitsTable) Partitions(*sql.Context) (sql.PartitionIter, error) {
...
+func (dt *CommitsTable) GetIndexes(ctx *sql.Context) ([]sql.Index, error) {
+	return index.DoltCommitIndexes(dt.Name(), dt.ddb, true)
+}
+
+// IndexedAccess implements sql.IndexAddressable
+func (dt *CommitsTable) IndexedAccess(lookup sql.IndexLookup) sql.IndexedTable {
+	if lookup.Index.ID() == index.CommitHashIndexId {
+		_, ok := index.LookupToCommitPointSelect(lookup)
+		if !ok {
+			return nil
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply go/libraries/doltcore/sqle/dtables/commits_table.go:95]: Type casting semantics, documented here, if the call returns nil we assume the indexing access is not supported. The analyzer aborts the attempted optimization. https://github.com/dolthub/go-mysql-server/pull/1540

File: go/libraries/doltcore/sqle/dtables/commits_table.go [hunk 5]
```diff
@@ -69,8 +73,84 @@ func (dt *CommitsTable) Partitions(*sql.Context) (sql.PartitionIter, error) {
...
+		var commits []*doltdb.Commit
+		var metas []*datas.CommitMeta
+		for _, hs := range hs {
+			h, ok := hash.MaybeParse(hs)
+			if !ok {
+				continue
+			}
+			hashes = append(hashes, h)
+
+			cm, err := doltdb.HashToCommit(ctx, dt.ddb.ValueReadWriter(), dt.ddb.NodeStore(), h)
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply go/libraries/doltcore/sqle/dtables/commits_table.go:118]: "WORKING" is the edge case that made this awkward. It is not always associated with a commit, and different system tables expect different behavior depending on whether it is a valid commit. I'll take another pass at trying to make it work with

File: go/libraries/doltcore/sqle/tables.go (hunk 3 above)
Comment (max-hoffman) [reply go/libraries/doltcore/sqle/tables.go:186]: the index was passed but not the range. so commit indexes are the first that don't support every range, or we didn't have enough tests to identify when other tables should be rejecting index lookups

File: go/libraries/doltcore/sqle/dtables/log_table.go [hunk 6]
```diff
@@ -90,8 +97,139 @@ func (dt *LogTable) Partitions(*sql.Context) (sql.PartitionIter, error) {
...
+}
+
+// CommitIsInScope returns true if a given commit hash is head or is
+// visible from the current head's ancestry graph.
+func (dt *LogTable) CommitIsInScope(ctx context.Context, height uint64, h hash.Hash) (bool, error) {
+	cc, err := dt.HeadCommitClosure(ctx)
+	if err != nil {
+		return false, err
+	}
+	headHeight, err := dt.HeadHeight()
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply go/libraries/doltcore/sqle/dtables/log_table.go:188]: tru

File: go/libraries/doltcore/doltdb/commit_itr.go [hunk 7]
```diff
@@ -191,3 +193,177 @@ func (itr FilteringCommitItr) Next(ctx context.Context) (hash.Hash, *Commit, err
 func (itr FilteringCommitItr) Reset(ctx context.Context) error {
 	return itr.itr.Reset(ctx)
 }
+
+func NewCommitIter(cm *Commit, m *datas.CommitMeta, h hash.Hash) (*CommitIter, error) {
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply go/libraries/doltcore/doltdb/commit_itr.go:197]: tried and hit import cycle with dtestutils

**Comment by max-hoffman** [comment]
#systab-benchmark

**Pull request merged**